hub/core/chunk_engine/read.py

```diff
@@ -35,27 +44,52 @@ def read_array(
         np.ndarray: Array containing the sample(s) in the `array_slice` slice.
     """
 
-    # TODO: don't use pickle
     meta = read_tensor_meta(key, storage)
-    index_map = pickle.loads(storage[get_index_map_key(key)])
+    index_map = read_index_map(key, storage)
 
     # TODO: read samples in parallel
     samples = []
     for index_entry in index_map[array_slice]:
-        chunks = []
-        for chunk_name in index_entry["chunk_names"]:
-            chunk_key = os.path.join(key, "chunks", chunk_name)
-            chunk = storage[chunk_key]
+        array = sample_from_index_entry(key, storage, index_entry, meta["dtype"])
+        samples.append(array)
 
-            chunks.append(chunk)
+    return np.array(samples)
 
-        combined_bytes = join_chunks(
-            chunks,
-            index_entry["start_byte"],
-            index_entry["end_byte"],
-        )
 
-        out_array = np.frombuffer(combined_bytes, dtype=meta["dtype"])
-        samples.append(out_array.reshape(index_entry["shape"]))
+def sample_from_index_entry(
+    key: str, storage: StorageProvider, index_entry: dict, dtype: str
+) -> np.ndarray:
+    """Get the unchunked sample from a single `index_map` entry."""
+
+    b = bytearray()
+    for chunk_name in index_entry["chunk_names"]:
+        chunk_key = os.path.join(key, "chunks", chunk_name)
+        last_b_len = len(b)
+        b.extend(storage[chunk_key])
+
+    start_byte = index_entry["start_byte"]
+    end_byte = last_b_len + index_entry["end_byte"]
+
+    return array_from_buffer(
+        memoryview(b),
+        dtype,
+        index_entry["shape"],
```

Comment:
Is this safe?

Reply:
not sure what you mean

Reply:
```
x = {}
x["shape"]
```
will throw an error. What I'm asking is, have we at some point asserted that `shape` is a member of `index_entry`?

Reply:
oh right, it is assumed to be because only arrays are supported as samples right now, but this will be implemented when we (@thisiseshan) implements the meta.json class